File: CPP/Easy/06Search2DMatrix.cpp
#include <bits/stdc++.h>
using namespace std;


bool searchMatrix(vector<vector<int>>& matrix, int x) {

        int n=matrix.size();
        int m=matrix[0].size();
        int low=0;
        int high=n*m-1;
        int mid=(low+high)/2;

        while(low<=high)
        {
            mid=(low+high)/2;

			if(matrix[mid/m][mid%m]<x)
            	low=mid+1;
            else if(matrix[mid/m][mid%m]==x)
            	return true;
            else
            	high=mid-1;
        }
        return false;
}


int main()
{
    vector<vector<int>> arr = {{1,2,3,4},
							{5,6,7,8},
							{9,10,11,12}};
    int target = 8;

    cout<<searchMatrix(arr, target)<<endl;

    return 0;
}

File: Python/Easy/06Search2DMatrix.py
# Using Binary Search Approach
matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]]
target = 3
def searchMatrix(matrix, target):
    for i in range(len(matrix)):
        low=0
        high=len(matrix[i])-1
        while low<=high:
            mid=low+(high-low)//2
            if matrix[i][mid]==target:
                return True
            elif matrix[i][mid]<target:
                low=mid+1
            else:
                high=mid-1
    return False

print(searchMatrix(matrix, target))